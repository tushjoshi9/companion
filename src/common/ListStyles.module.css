.listWrapper {
  background-color: gray;
}
.flexbox {
  display: flex;
}
.title {
  flex: 2;
  text-align: left;
  font-weight: 700;
  text-transform: uppercase;
  position: relative;
  padding-left: 32px;
  color: #434c5e;
}
.header {
  align-items: center;
  padding: 10px 0 20px;
  border-bottom: 1px #e9e9eb solid;
  margin-bottom: 20px;
}
.list {
  padding: 0;
  margin: 0;
  margin-top: 20px;
  height: 380px;
}
.list li {
  padding: 20px;
  border: 1px #e9e9eb solid;
  margin: 10px 0;
  border-radius: 2px;
  box-shadow: 0 0 10px 1px #ddd;
  display: flex;
  align-items: center;
  justify-content: space-between;
}
.searchBox {
  padding: 10px 20px 10px 38px;
  width: 100%;
  border: 1px #e9e9eb solid;
  border-radius: 40px;
  box-sizing: border-box;
  font-size: 13px;
  flex: 2;
  background: url("../assets/search.png") no-repeat center left 15px/16px;
}
.addFriend {
  padding: 15px 20px;
  width: 100%;
  border: 1px #e9e9eb solid;
  border-radius: 2px;
  font-size: 16px;
  box-sizing: border-box;
  background: url("../assets/add.png") no-repeat center right 10px/20px;
}

.listtray {
  align-items: center;
}
.friendName {
  text-align: left;
  font-weight: 600;
}
.friendNote {
  font-size: 11px;
  font: italic 400 12px "Inter", sans-serif;
  color: #8a929e;
  margin-top: 2px;
}
.actions {
  background: url("../assets/star-outline.png") no-repeat center center/20px;
  width: 20px;
  height: 20px;
  border: 0;
  outline: none;
  margin-left: 15px;
  cursor: pointer;
}
.actionWrapper {
  display: flex;
}

.favourite[data-fav="true"] {
  background: url("../assets/star-filled.png") no-repeat center center/20px;
}
.remove {
  background: url("../assets/remove.png") no-repeat center center/20px;
  width: 20px;
  height: 20px;
}

.activePage {
  color: #ffffff;
  background-color: #28bd8b;
  padding: 5px 10px;
  border-radius: 2px;
  cursor: pointer;
}
.inActivePage {
  color: #434c5e;
  padding: 0 10px;
  cursor: pointer;
}
.searchResult {
  margin-top: 20px;
  font: 400 20px "Inter", sans-serif;
  color: #434c5e;
}
.removeButton {
  font: 400 13px "Inter", sans-serif;
  padding: 10px 20px;
  background-color: #468def;
  border-radius: 2px;
  color: #ffffff;
  border: none;
  cursor: pointer;
}
