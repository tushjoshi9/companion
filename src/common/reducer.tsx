import { friendState } from "../friends-list"

type Action =
  | { type: 'addFriend'; payload: string; }
  | { type: 'nextPage'; payload: number; }
  | { type: 'searchFriend'; payload: string; }
  | { type: 'removeFriend'; payload: number; }
  | { type: 'favouriteFriend'; payload: { id: number; value: boolean; }; }

export function friendsReducer(state: friendState, action: Action) {
  switch (action.type) {
    case "addFriend":
      return {
        ...state,
        friendList: [...state.friendList, { name: action.payload, id: state.friendList.length, favourite: false }]
      };
    case "removeFriend":
      return {
        ...state,
        friendList: state.friendList.filter((friend) => friend.id !== action.payload),
        searchFriendList: state.searchFriendList.filter((friend) => friend.id !== action.payload),
      };
    case "favouriteFriend": {
      let friendList = state.friendList;
      let getIndex = friendList.findIndex(friend => friend.id === action.payload.id);
      friendList[getIndex].favourite = action.payload.value;
      friendList = friendList.sort((a, b) => Number(b.favourite) - Number(a.favourite));
      return { ...state, friendList };
    };
    case "searchFriend":
      return action.payload ?
        {
          ...state,
          searchFriendList: state.friendList.filter(friend => friend.name?.toLowerCase().includes(action.payload.toLowerCase()))
        }
        : state;
    case "nextPage":
      return {
        ...state,
        currentPage: action.payload
      };
    default:
      throw new Error();
  }
}