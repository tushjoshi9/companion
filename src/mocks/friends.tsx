export const mockFriends = {
  friendList: [
    { id: 0, name: "Suraj", favourite: false },
    { id: 1, name: "Dheeraj", favourite: false }
  ],
  searchFriendList: [],
  currentPage: 1,
  todosPerPage: 4
}