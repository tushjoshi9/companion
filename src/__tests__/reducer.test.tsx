import { useReducer } from 'react';
import { act, renderHook } from "@testing-library/react-hooks";
import { friendsReducer } from "../common/reducer";
import { mockFriends } from "../mocks/friends";

describe("Friends reducer", () => {
  test("should add a new friend", () => {
    const { result } = renderHook(() => useReducer(friendsReducer, mockFriends));
    const [, dispatch] = result.current;
    act(() => {
      dispatch({ type: "addFriend", payload: "Test Name" });
    })
    expect(result.current[0]).toMatchSnapshot();
  });

  test("should remove a friend", () => {
    const { result } = renderHook(() => useReducer(friendsReducer, mockFriends));
    const [, dispatch] = result.current;
    act(() => {
      dispatch({ type: "removeFriend", payload: 1 });
    })
    expect(result.current[0]).toMatchSnapshot();
  });

  test("should favourite a friend", () => {
    const { result } = renderHook(() => useReducer(friendsReducer, mockFriends));
    const [, dispatch] = result.current;
    act(() => {
      dispatch({
        type: "favouriteFriend", payload: {
          id: 2,
          value: true
        }
      });
    })
    expect(result.current[0]).toMatchSnapshot();
  });

  test("should show nextpage", () => {
    const { result } = renderHook(() => useReducer(friendsReducer, mockFriends));
    const [, dispatch] = result.current;
    act(() => {
      dispatch({
        type: "nextPage", payload: 2
      });
    })
    expect(result.current[0]).toMatchSnapshot();
  });

  test("should search friends list", () => {
    const { result } = renderHook(() => useReducer(friendsReducer, mockFriends));
    const [, dispatch] = result.current;
    act(() => {
      dispatch({
        type: "searchFriend", payload: "DJ"
      });
    })
    expect(result.current[0]).toMatchSnapshot();
  });
})
