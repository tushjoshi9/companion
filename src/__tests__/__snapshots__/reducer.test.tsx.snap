// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends reducer should add a new friend 1`] = `
Object {
  "currentPage": 0,
  "friendList": Array [
    Object {
      "favourite": false,
      "id": 0,
      "name": "Suraj",
    },
    Object {
      "favourite": false,
      "id": 1,
      "name": "Dheeraj",
    },
    Object {
      "favourite": false,
      "id": 2,
      "name": "Test Name",
    },
  ],
  "searchFriendList": Array [],
  "todosPerPage": 4,
}
`;

exports[`Friends reducer should favourite a friend 1`] = `
Object {
  "currentPage": 0,
  "friendList": Array [
    Object {
      "favourite": true,
      "id": 1,
      "name": "Dheeraj",
    },
    Object {
      "favourite": false,
      "id": 0,
      "name": "Suraj",
    },
  ],
  "searchFriendList": Array [],
  "todosPerPage": 4,
}
`;

exports[`Friends reducer should remove a friend 1`] = `
Object {
  "currentPage": 0,
  "friendList": Array [
    Object {
      "favourite": false,
      "id": 0,
      "name": "Suraj",
    },
  ],
  "searchFriendList": Array [],
  "todosPerPage": 4,
}
`;

exports[`Friends reducer should search friends list 1`] = `
Object {
  "currentPage": 0,
  "friendList": Array [
    Object {
      "favourite": true,
      "id": 1,
      "name": "Dheeraj",
    },
    Object {
      "favourite": false,
      "id": 0,
      "name": "Suraj",
    },
  ],
  "searchFriendList": Array [],
  "todosPerPage": 4,
}
`;

exports[`Friends reducer should show nextpage 1`] = `
Object {
  "currentPage": 2,
  "friendList": Array [
    Object {
      "favourite": true,
      "id": 1,
      "name": "Dheeraj",
    },
    Object {
      "favourite": false,
      "id": 0,
      "name": "Suraj",
    },
  ],
  "searchFriendList": Array [],
  "todosPerPage": 4,
}
`;
